src/library/browse/foldertreemodel.cpp

```diff
@@ -1,27 +1,99 @@
+        : TreeItemModel(parent), m_isRunning(true) {
+    connect(&m_fsWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &FolderTreeModel::dirModified);
+    connect(&m_fsWatcher,
+            &QFileSystemWatcher::directoryChanged,
+            this,
+            &FolderTreeModel::dirModified);
+    connect(this,
+            &FolderTreeModel::newChildren,
+            this,
+            &FolderTreeModel::addChildren,
+            Qt::QueuedConnection);
+    connect(this,
+            &FolderTreeModel::hasSubDirectory,
+            this,
+            &FolderTreeModel::onHasSubDirectory);
+
+    m_pool.setMaxThreadCount(4);
+    m_folderProcess = QtConcurrent::run(&m_pool, [&]() {
+        while (m_isRunning.load(std::memory_order_consume)) {
+            if (!m_folderQueue.isEmpty()) {
+                m_queueLock.lock();
+                const auto& [parent, path] = m_folderQueue.dequeue();
+                m_queueLock.unlock();
 
-FolderTreeModel::FolderTreeModel(QObject *parent)
-        : TreeItemModel(parent) {
+                m_cacheLock.lockForWrite();
+                const auto dirAccess = mixxx::FileAccess(mixxx::FileInfo(path));
+                QFileInfoList all = dirAccess.info().toQDir().entryInfoList(
+                        QDir::Dirs | QDir::NoDotAndDotDot);
+
+                auto* folders = new QList<TreeItem*>();
+                QFutureSynchronizer<void> sync;
+                // loop through all the item and construct the children
+                foreach (QFileInfo one, all) {
+#if defined(__APPLE__)
+                    if (one.isDir() && one.fileName().endsWith(".app")) {
+                        continue;
+                    }
+#endif
+                    // We here create new items for the sidebar models
+                    // Once the items are added to the TreeItemModel,
+                    // the models takes ownership of them and ensures their deletion
+                    const auto& absPath = one.absoluteFilePath();
+                    auto* folder =
+                            new TreeItem(one.fileName(), QVariant(absPath));
+                    // init cache
+                    if (m_directoryCache.find(absPath) ==
+                            m_directoryCache.end()) {
+                        m_fsWatcher.addPath(absPath);
+                        sync.addFuture(
+                                QtConcurrent::run(&m_pool, [this, absPath]() {
+                                    this->directoryHasChildren(absPath);
+                                }));
+                    }
+                    folders->push_back(folder);
+                }
+                if (!folders->isEmpty()) {
+                    sync.waitForFinished();
+                    m_cacheLock.unlock();
+                    emit newChildren(parent, folders);
+                } else {
+                    m_cacheLock.unlock();
+                }
+            }
+            QThread::msleep(100);
```

Comment:
Busy looking at the `m_folderQueue` is wasting CPU. Can we replace it with a single Future that is started for every folder added? This was we would get rid of the m_folderQueue altogether. This would also solve the issue with the signal from the Future. It would be just replaced with a `QFuture::result()`